kubetest/aks.go

```diff
@@ -0,0 +1,297 @@
+	"os/exec"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/services/containerservice/mgmt/2019-10-01/containerservice"
+	"github.com/Azure/go-autorest/autorest/azure"
+	uuid "github.com/satori/go.uuid"
+	"golang.org/x/crypto/ssh"
+)
+
+type aksDeployer struct {
+	azureCreds    *Creds
+	azureClient   *AzureClient
+	templateUrl   string
+	outputDir     string
+	resourceGroup string
+	resourceName  string
+}
+
+func newAksDeployer() (*aksDeployer, error) {
+	if err := validateAksFlags(); err != nil {
+		return nil, err
+	}
+
+	creds, err := getAzCredentials()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get azure credentials: %v", err)
+	}
+	env, err := azure.EnvironmentFromName(*aksAzureEnv)
+	if err != nil {
+		return nil, fmt.Errorf("failed to determine azure environment: %v", err)
+	}
+
+	var client *AzureClient
+	if client, err = getAzureClient(env,
+		creds.SubscriptionID,
+		creds.ClientID,
+		creds.TenantID,
+		creds.ClientSecret,
+	); err != nil {
+		return nil, fmt.Errorf("error trying to get Azure Client: %v", err)
+	}
+
+	tempdir, err := ioutil.TempDir(os.Getenv("HOME"), "aks")
+	if err != nil {
+		return nil, fmt.Errorf("error creating tempdir: %v", err)
+	}
+
+	return &aksDeployer{
+		azureCreds:    creds,
+		azureClient:   client,
+		templateUrl:   *aksTemplateURL,
+		outputDir:     tempdir,
+		resourceGroup: *aksResourceGroupName,
+		resourceName:  *aksResourceName,
+	}, nil
+}
+
+func validateAksFlags() error {
+	if *aksCredentialsFile == "" {
+		return fmt.Errorf("no credentials file path specified")
+	}
+	if *aksResourceName == "" {
+		*aksResourceName = "kubetest-" + uuid.NewV1().String()
+	}
+	if *aksResourceGroupName == "" {
+		*aksResourceGroupName = *aksResourceName
+	}
+	if *aksDNSPrefix == "" {
+		*aksDNSPrefix = *aksResourceName
+	}
+	return nil
+}
+
+func (a *aksDeployer) Up() error {
+	log.Printf("Creating AKS cluster %v in resource group %v", a.resourceName, a.resourceGroup)
+	templateFile, err := downloadFromURL(a.templateUrl, path.Join(a.outputDir, "kubernetes.json"), 2)
+	if err != nil {
+		return fmt.Errorf("error downloading AKS cluster template: %v with error %v", a.templateUrl, err)
+	}
+
+	template, err := ioutil.ReadFile(templateFile)
+	if err != nil {
+		return fmt.Errorf("failed to read downloaded cluster template file: %v", err)
+	}
+
+	var model containerservice.ManagedCluster
+	if err := json.Unmarshal(template, &model); err != nil {
+		return fmt.Errorf("failed to unmarshal managedcluster model: %v", err)
+	}
+
+	_, sshPublicKey, err := newSSHKeypair(4096)
+	if err != nil {
+		return fmt.Errorf("failed to generate ssh key for cluster creation: %v", err)
+	}
+
+	*(*model.LinuxProfile.SSH.PublicKeys)[0].KeyData = string(sshPublicKey)
+	model.ManagedClusterProperties.DNSPrefix = aksDNSPrefix
+	model.ManagedClusterProperties.ServicePrincipalProfile.ClientID = &a.azureCreds.ClientID
+	model.ManagedClusterProperties.ServicePrincipalProfile.Secret = &a.azureCreds.ClientSecret
```

Comment:
I think it's better to also set region and kubernetes version here

Reply:
fixed, parsing from the shared flags with aks-e currently